adapter/src/hermes/adapter/stdio/metadata_manager.h

```diff
@@ -0,0 +1,115 @@
+#ifndef HERMES_ADAPTER_METADATA_MANAGER_H
+#define HERMES_ADAPTER_METADATA_MANAGER_H
+
+/**
+ * Standard headers
+ */
+#include <ftw.h>
+
+#include <cstdio>
+#include <unordered_map>
+
+/**
+ * Internal headers
+ */
+#include <hermes/adapter/stdio/common/constants.h>
+#include <hermes/adapter/stdio/common/datastructures.h>
+
+namespace hermes::adapter::stdio {
+/**
+ * Metadata manager for STDIO adapter
+ */
+class MetadataManager {
+ private:
+  /**
+   * Private members
+   */
+  /**
+   * Maintain a local metadata FileID structure mapped to Adapter Stats.
+   */
+  std::unordered_map<FileID, AdapterStat> metadata;
+  /**
+   * hermes attribute to initialize Hermes
+   */
+  std::shared_ptr<hapi::Hermes> hermes;
+  /**
+   * references of how many times hermes was tried to initialize.
+   */
+  size_t ref;
+
+ public:
+  /**
+   * Constructor
+   */
+  MetadataManager() : metadata(), ref(0) {}
+  /**
+   * Get the instance of hermes.
+   */
+  std::shared_ptr<hapi::Hermes>& GetHermes() { return hermes; }
+
+  /**
+   * Initialize hermes. Get the kHermesConf from environment else get_env
+   * returns NULL which is handled internally by hermes. Initialize hermes in
+   * daemon mode. Keep a reference of how many times Initialize is called.
+   * Within the adapter, Initialize is called from fopen.
```

Comment:
Is there a reason you didn't use the `_init` mechanism to initialize Hermes once when the adapter `.so` is loaded?

Reply:
U mean MPI init?

Reply:
No. The `-init` flag to `ld` lets you specify a function that is called when the shared library loads. That way you won't have to check for Hermes initialization in every `fopen` or keep a refcount.

Reply:
Do u have a reference i can follow.

Reply:
There's a descent summary [here](https://stackoverflow.com/a/9759936/2059743). It looks like `__attribute__(constructor)` is preferable to `ld -init` these days. This is a GCC extension though, so if people are using other compilers on the machines we're targeting, maybe we should leave it as is.

Reply:
Yes that is what i read. I will explore this option and let u know.,

Reply:
I0121 12:15:49.370560  5749 memory_management.cc:128] PushSize added 2 bytes of padding for alignment
I0121 12:15:49.373865  5749 memory_management.cc:128] PushSize added 2 bytes of padding for alignment
Attempting to use an MPI routine before initializing MPICH

The constructor is called before main so MPI is not initialized. I think this approach won't work for us. We should do the init on fopen.